### PR TITLE
Overhaul how cargo treats profiles

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is a complete overhaul of how Cargo handles compilation profiles
internally. The external interface of Cargo is not affected by this change.

Previously each Target had a Profile embedded within it. Conceptually a Target
is an entry in the manifest (a binary, benchmark, etc) and a Profile controlled
various flags (e.g. --test, -C opt-level, etc). Each Package then contained many
profiles for each possible compilation mode. For example a Package would have
one target for testing a library, benchmarking a library, documenting a library,
etc. When it came to building these targets, Cargo would filter out the targets
listed to determine what needed to be built.

This filtering was largely done based off an "environment" represented as a
string. Each mode of compilation got a separate environment string like `"test"`
or `"bench"`. Altogether, however, this approach had a number of drawbacks:

* Examples in release mode do not currently work. This is due to how examples
  are classified and how release mode is handled (e.g. the "release" environment
  where examples are meant to be built in the "test" environment).
* It is not trivial to implement `cargo test --release` today.
* It is not trivial to implement `cargo build --bin foo` where *only* the binary
  `foo` is built. The same is true for examples.
* It is not trivial to add selective building of a number of
  binaries/examples/etc.
* Filtering the list of targets to build has some pretty hokey logic that
  involves pseudo-environments like "doc-all" vs "doc". This logic is duplicated
  in a few places and in general is quite brittle.
* The TOML parser greatly affects compilation due to the time at which profiles
  are generated, which seems somewhat backwards.
* Profiles must be overridden, but only partially, at compile time becuase only
  the top-level package's profile is applied.

In general, this system just needed an overhaul. This commit made a single
change of separating `Profile` from `Target` and then basically hit `make` until
all the tests passed again. The other large architectural changes are:

* Environment strings are now entirely gone.
* Filters are implemented in a much more robust fashion.
* Release mode is now handled much more gracefully.
* The unit of compilation in the backend is no longer (package, target) but
  rather (package, target, profile). This change had to be propagated many
  location in the `cargo_rustc` module.
* The backend does not filter targets to build *at all*. All filtering now
  happens entirely in the frontend.

I'll test issues after this change lands, but the motivation behind this is to
open the door to quickly fixing a number of outstanding issues against Cargo.
This change itself is not intended to close many bugs.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This is me finally following up to #1325

**Comment by alexcrichton** [comment]
cc @wycats

**Comment by alexcrichton** [comment]
This will also unblock some work related to rust-lang/rust#23533 I've got pending.

File: src/bin/bench.rs [hunk 1]
```diff
@@ -50,20 +50,31 @@ pub fn execute(options: Options, config: &Config) -> CliResult<Option<()>> {
...
             target: options.flag_target.as_ref().map(|s| &s[..]),
-            dev_deps: true,
             features: &options.flag_features,
             no_default_features: options.flag_no_default_features,
             spec: options.flag_package.as_ref().map(|s| &s[..]),
-            lib_only: false,
             exec_engine: None,
+            release: true,
+            mode: ops::CompileMode::Bench,
+            filter: if benches.len() == 0 {
```
Comment (tamird) [inline src/bin/bench.rs]: nit: `is_empty()`

File: src/bin/bench.rs [hunk 2]
```diff
@@ -50,20 +50,31 @@ pub fn execute(options: Options, config: &Config) -> CliResult<Option<()>> {
     let root = try!(find_root_manifest_for_cwd(options.flag_manifest_path));
     config.shell().set_verbose(options.flag_verbose);
 
+    let mut benches = Vec::new();
```
Comment (tamird) [inline src/bin/bench.rs]: this vector looks unnecessary - seems you could use `options.flag_bench.as_slice()`

File: src/bin/run.rs [hunk 3]
```diff
@@ -47,36 +46,35 @@ pub fn execute(options: Options, config: &Config) -> CliResult<Option<()>> {
     config.shell().set_verbose(options.flag_verbose);
     let root = try!(find_root_manifest_for_cwd(options.flag_manifest_path));
 
-    let env = match (options.flag_release, options.flag_example.is_some()) {
-        (true, _) => "release",
-        (false, true) => "test",
-        (false, false) => "compile"
-    };
+    let (mut examples, mut bins) = (Vec::new(), Vec::new());
```
Comment (tamird) [inline src/bin/run.rs]: ditto singleton vectors

File: src/bin/test.rs [hunk 4]
```diff
@@ -52,20 +54,34 @@ pub fn execute(options: Options, config: &Config) -> CliResult<Option<()>> {
     let root = try!(find_root_manifest_for_cwd(options.flag_manifest_path));
     config.shell().set_verbose(options.flag_verbose);
 
+    let (mut tests, mut bins) = (Vec::new(), Vec::new());
```
Comment (tamird) [inline src/bin/test.rs]: and here

File: src/bin/bench.rs (hunk 2 above)
Comment (alexcrichton) [inline src/bin/bench.rs]: I'm currently trying to keep Cargo on stable Rust as much as possible, and unfortunately these methods are currently unstable. I had not considered this though, and it's a good idea! I may try to facilitate the stability of `as_slice` now...

**Comment by alexcrichton** [comment]
I can also squash any number of commits as desired or break up the first commit if necessary. I started by lumping as much as possible into one commit but eventually started making one fix per mistake, so the first commit is somewhat biasedly large!

**Pull request merged**